src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,49 +1,111 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
 
+    private static final int DEFAULT_CAPACITY = 10;
+    private static final double RESIZE_COEFFICIENT = 1.5;
+    private int size;
+    private T[] values;
+
+    public ArrayList() {
+        this.values = (T[]) new Object[DEFAULT_CAPACITY];
+        this.size = 0;
+    }
+
+    public ArrayList(int initialCapacity) {
+        this.values = (T[]) new Object[initialCapacity];
+        this.size = 0;
+    }
+
+    private void resize() {
+        T[] resizedValues = (T[]) new Object[(int) Math.ceil(values.length * RESIZE_COEFFICIENT)];
+        System.arraycopy(this.values, 0, resizedValues, 0, this.values.length);
+        this.values = resizedValues;
+    }
+
+    private void isValidIndex(int index) {
+        if (index >= this.size) {
+            throw new ArrayIndexOutOfBoundsException(
+                    "Index " + index + " out of bounds for capacity " + this.values.length);
+        }
+    }
+
     @Override
     public void add(T value) {
-
+        add(value, size);
     }
 
     @Override
     public void add(T value, int index) {
-
+        if (index == size) {
+            if (size == this.values.length) {
+                resize();
+            }
+            size++;
+            values[index] = value;
+        } else if (index <= size - 1) {
+            resize();
+            size++;
+            System.arraycopy(
+                    this.values, index, this.values, index + 1, this.values.length - index - 1);
+            values[index] = value;
+        }
+        isValidIndex(index);
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        for (int i = 0; i < list.size(); i++) {
+            this.add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        isValidIndex(index);
+        return (T) values[index];
     }
 
     @Override
     public void set(T value, int index) {
-
+        isValidIndex(index);
+        values[index] = value;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        isValidIndex(index);
+        if (index == this.size - 1) {
+            size--;
+            return (T) values[index];
+        } else {
+            Object item = values[index];
+            System.arraycopy(
+                    this.values, index + 1, this.values, index, this.values.length - index - 1);
+            size--;
+            return (T) item;
+        }
     }
 
     @Override
     public T remove(T t) {
-        return null;
+        for (int i = 0; i < this.size; i++) {
+            if (values[i] == null || values[i].equals(t)) {
```

Comment:
is this the right way to compare two objects?
what if both of them are null?

Reply:
If `t` is`null` -> if `values[i]` is null, we remove right away; if `values[i]` is not null, we continue.
if `t` is not `null` -> if `values[i]` is null, we continue; if `values[i]` is not null, we compare by `.equals()`

Reply:
> if t is not null -> if values[i] is null, we continue

we won't continue - the left part of the || will be true

Reply:
Damn, these null values...